### DISCUSSION TITLE
Should SetMouseCursor method add a Context pointer parameter?

### DISCUSSION THREAD

**Discussion by xland** [mikke89/RmlUi#403] [Q&A] [answered]
When an application has multiple windows.
`SetMouseCursor` method in backend's `SystemInterface` doesn't know which window's context to be updated.
In addition, `SetClipboardText` and `GetClipboardText` methods have the same problem.

```c++
//Core/Context.cpp
//......
if(enable_cursor)
{
    String new_cursor_name;    
    if(drag)
	    new_cursor_name = drag->GetComputedValues().cursor();
    else if (hover)
	    new_cursor_name = hover->GetComputedValues().cursor();    
    if(new_cursor_name != cursor_name)
    {
	    GetSystemInterface()->SetMouseCursor(new_cursor_name);   //here
	    cursor_name = new_cursor_name;
    }
}
//.....
```

**Comment by mikke89** [comment, marked as answer]
It is generally assumed that there is only one mouse cursor and only one clipboard. If this is the case, it shouldn't really matter which context it is set from. Otherwise, what exactly is the use case you need this for?

Perhaps you can make use of `Context::EnableMouseCursor` so that only a single context will transmit the mouse cursor to the system interface at a time, and you can know where it came from that way.

**Reply by xland** [reply]
I'm trying to add another window to my application which use `RmlUi_Backend_GLFW_GL3` backend.
This backend only support one window per application.
I think I must make a backend by myself to support multiple windows.
But I found that I could not achieve `SetMouseCursor` method

```c++
// RmlUi_Platform_GLFW.cpp
void SystemInterface_GLFW::SetMouseCursor(const Rml::String& cursor_name)
{
	GLFWcursor* cursor = nullptr;
	if (cursor_name.empty() || cursor_name == "arrow")
		cursor = nullptr;
	else if (cursor_name == "move")
		cursor = cursor_pointer;
	else if (cursor_name == "unavailable")
		cursor = nullptr;
	if (window)
		glfwSetCursor(window, cursor);  //here
}
```

`glfwSetCursor(window, cursor);` need the window pointer.
I cann't get the window pointer by `cursor_name`.

Even if the system interface instance holds multiple window pointers, 
I don't know which window pointer the `SetMouseCursor` method needs.

`EnableMouseCursor` can't solve this problem

**Reply by xland** [reply]
`RmlUi_Platform_SFML` and `RmlUi_Platform_Win32` have the same problem.

**Reply by xland** [reply]
@mikke89 
I am making a calendar project with RmlUi.
When users want to add a to-do item to the calendar, they need to open a modal window.
At present, I am struggling on how to create a modal window
![image](https://user-images.githubusercontent.com/2980915/212850431-10b20bd3-c269-47f5-99b4-d0761efa8a17.png)
This is an open source project. 
The project address is https://github.com/xland/51richeng/tree/RmlUi2

**Reply by xland** [reply]
![image](https://user-images.githubusercontent.com/2980915/212851094-9c1fb75c-6225-4e53-a1e0-bc80b2d9251b.png)

**Reply by mikke89** [reply]
I can think of at least a few ways you can solve this without us modifying the API.

1. You could keep track of which window the mouse is hovering over. See the glfw cursor enter/leave guide: https://www.glfw.org/docs/3.3/input_guide.html#cursor_enter
2. Another option is to store a pointer to the context manually in your system interface, setting it every time before you call into the context.
3. Update the cursor on all the windows simultaneously.

By the way, I see that the clipboard window parameter is deprecated in glfw, you can just pass NULL to it.

The application you're making looks really smooth by the way! I'd be happy to add it to the gallery.

**Reply by xland** [reply]
Thanks very much! I will try your suggestion.
